### ISSUE TITLE
Repository Watcher does not work on non-JPA Repositories

### ISSUE THREAD

**Issue by ericwyles** [codecentric/chaos-monkey-spring-boot#147]
## Expected Behavior
Chaos monkey should be able to assault @Repository classes that are not implementations of CrudRepository or within RepositoryDefinition. We use spring-data-jdbc and have several classes annotated with @Repository. This gives us exception translation and other benefits from the repository layer, but these are not JPA repositories.

## Current Behavior
The SpringRepositoryAspect pointcut only picks up CrudRepository or RepositoryDefinition classes. Classes annotated with @Repository are not considered.

## Possible Solution
Modify SpringRepositoryAspect to allow @Repository classes to be picked up by the pointcut.

## Steps to Reproduce
Make a simple class annotated with @Repository and try to watch it and assault it with Chaos monkey. It will not work.

**Comment by ericwyles** [comment]
I submitted a PR with a fix for this. I created an entire new aspect to reference the Repository stereotype and updated the auto configure to create that bean even when the CrudRepository is not on the classpath. 

#148 

I did it this way because trying to combine it with the existing aspect was causing issues in apps that don't have CrudRepository on the classpath. They would compile but not be able to boot because the Pointcut expression referenced a class that did not exist. Keeping it as two separate aspects allows those apps towork without adding spring-data related classes to the classpath just for chaos monkey.

**Comment by maiksensi** [comment]
Hey @ericwyles,

thanks for your MR! It is highly appreciated. However, I am not sure we will get to it before Monday. I hope that is fine for you.

**Comment by ericwyles** [comment]
@maiksensi That's totally fine with me. I hope the approach I took works out ok and fits with the project. I enabled edits for project maintainers on my source branch as well, so if you need to make any tweaks feel free.

**Comment by maiksensi** [comment]
I had a brief look at things and I was wondering why the current implementation watches for `CrudRepository` anyways. I would have expected the Pointcut in the same way as the ones for controllers or services (or the one you added). They basically watch for the appropriate annotation. 

Thus we should be able to remove 

```
    @Bean
    @DependsOn("chaosMonkeyRequestScope")
    // Don't create this bean if Spring Data is not on the classpath.
    @ConditionalOnClass(name = "org.springframework.data.repository.CrudRepository")
    public SpringRepositoryAspect repositoryAspect(ChaosMonkeyRequestScope chaosMonkeyRequestScope) {
        return new SpringRepositoryAspect(chaosMonkeyRequestScope, publisher(), watcherProperties);
    }
```
and all corresponding usages. `SpringRepositoryAspectTest` should be able to stay the same but we should add your test case to it (`DemoRepositoryStereotype`).
I haven't had the time to test that assumption yet. If possible, I'd like to have a clean solution here. 

@WtfJoke & @fletchgqc any opinions on that?

**Comment by ericwyles** [comment]
I agree that would be cleaner and that was my original intent. I actually implemented it that way first.

The problem I ran into was that in the typical CrudRepository implementations the @Repository annotation is on the interface and not on the implementation. Making that change broke the existing test cases because the repository implementations were no longer being picked up by the aspect.

I stopped at that point and changed directions because I was concerned that would be a big breaking change and I was also not sure if it would even work in cases where Spring Data automatically generates implementations from the interface.

If we can get that design working properly for all the cases I completely agree it would be a better solution and a lot cleaner.

**Comment by ericwyles** [comment]
There was actually one other issue. I also tried just adding @Repository to the pointcut in the existing aspect, and removing ConditionalOnClass from the bean creation.

The issue with that was the pointcut evaluation still requires CrudRepository.class to be on the classpath or else the apps won't start. So an app that uses spring-data-jdbc classes but nothing from spring-data-commons will still have to add a dependency to spring-data-commons just to get the Pointcut evaluation to not throw an exception when it checks for CrudRepository.

**Comment by maiksensi** [comment]
Hey @ericwyles,

I've had the time to look into this a bit more. Thanks for your explanations, they helped a great deal to track down the problem 👍.
I am pretty sure we get around the need to add another separate aspect by doing the following in `SpringRepositoryAspect`: 

```
    @Pointcut("this(org.springframework.data.repository.Repository) || within(@org.springframework.data.repository.RepositoryDefinition *)")
        public void isRepositoryImplementation() {
    }

    @Pointcut("within(@org.springframework.stereotype.Repository *)")
    public void isRepositoryAnnotation() {
    }

    @Around("(isRepositoryAnnotation() || isRepositoryImplementation())) && allPublicMethodPointcut() && !classInChaosMonkeyPackage()")

```
The idea is, we either wrap around spring-data implementations or classes annotated with `@Repository` (`(isRepositoryAnnotation() || isRepositoryImplementation()))`).
Notes:
- Changed naming to be more speaking  
- Removed the original Pointcut 
- I would not match for `CrudRepository` anymore but instead match for `Repository`, which is the base class (of `CrudRepository`). This way we can also match `PagingAndSortingRepository` and so on. 

🚀 Thanks to your contribution we figured out that we should have a look at how we compose the aspects. I talked to @WtfJoke and we agreed on thinking about it in a later refactoring. 

I know you granted edit rights to your fork, but we would love to see the changes coming from you. Give the suggested changes a test run within your context and let us know if that works for you. If the suggested changes work out, please don't forget to add your contribution to the changelog! 🎉

(optional) If you want to add some notes into the docs that you we either match classes annotated with `@Repository` or implementations of spring-data or custom implementations extending `Repository` or children classes of it. 

(optional) The new `@Around` looks a bit different now. An explanation could help us remember what we did in the future 😃 .  

Please let me know if that makes sense and works for you.

**Comment by fletchgqc** [comment]
I added the comment "// Don't create this bean if Spring Data is not on the classpath." and can confirm that @ericwyles' explanation is correct, this is why I added it. I did not however add the CrudRepository, and I feel that the design is incorrect to focus on this one implementation. There are various types of repositories that can be used, as you are basically saying, and I think that they should be controlled in a unified way by one watcher. Otherwise you increase complication for users. So I like the concept behind the "clean implementation" you are aiming for.

**Comment by ericwyles** [comment]
@maiksensi Yes, that all makes sense. I'll make the updates.

**Comment by fletchgqc** [comment]
@maiksensi are you sure that matching against base classes works? I thought that it specifically does not work with pointcuts.

isRepositoryAnnotation should be hasRepositoryAnnotation.

**Comment by fletchgqc** [comment]
I take that back, I tested it by putting your pointcut suggestions into SpringRepositoryAspectTest and it works. Ok, over to @ericwyles .

**Comment by ericwyles** [comment]
@maiksensi I made the suggested changes but even with the new Pointcuts in place, when I remove the   @ConditionalOnClass(name = "org.springframework.data.repository.CrudRepository") from ChaosMonkeyConfiguration, it causes problems for any app that doesn't use spring-data-commons

Those apps fail to start with this error:

```
Caused by: java.lang.IllegalArgumentException: warning no match for this type name: org.springframework.data.repository.Repository [Xlint:invalidAbsoluteTypeName]
	at org.aspectj.weaver.tools.PointcutParser.parsePointcutExpression(PointcutParser.java:319) ~[aspectjweaver-1.9.4.jar:na]
	at org.springframework.aop.aspectj.AspectJExpressionPointcut.buildPointcutExpression(AspectJExpressionPointcut.java:227) ~[spring-aop-5.2.0.RELEASE.jar:5.2.0.RELEASE]
	at org.springframework.aop.aspectj.AspectJExpressionPointcut.obtainPointcutExpression(AspectJExpressionPointcut.java:198) ~[spring-aop-5.2.0.RELEASE.jar:5.2.0.RELEASE]
	at org.springframework.aop.aspectj.AspectJExpressionPointcut.getClassFilter(AspectJExpressionPointcut.java:177) ~[spring-aop-5.2.0.RELEASE.jar:5.2.0.RELEASE]
	at org.springframework.aop.support.AopUtils.canApply(AopUtils.java:226) ~[spring-aop-5.2.0.RELEASE.jar:5.2.0.RELEASE]
	at org.springframework.aop.support.AopUtils.canApply(AopUtils.java:289) ~[spring-aop-5.2.0.RELEASE.jar:5.2.0.RELEASE]
	at org.springframework.aop.support.AopUtils.findAdvisorsThatCanApply(AopUtils.java:321) ~[spring-aop-5.2.0.RELEASE.jar:5.2.0.RELEASE]
```

This is the reason I originally made the seperate aspect, because it's safe to check for the repository annotation in any spring boot app, but even checking for this Repository interface requires spring-data specific classes on the classpath. I was hopeful that the new Pointcut would avoid the problem but it's still there.

Any other ideas?

**Comment by maiksensi** [comment]
Hey @ericwyles,

hmm okay. We tested it with one of the demo apps:  https://github.com/codecentric/chaos-monkey-spring-boot/tree/master/demo-apps/chaos-monkey-web-reactive-app 

Can you please check if the changes you applied work with that app? Running `mvn test` should do the trick and also start that app.

**Comment by ericwyles** [comment]
mvn test on chaos-monkey-web-reactive-app gives the same error:

```
  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::        (v2.1.8.RELEASE)

2019-11-13 14:08:49.870  INFO 16448 --- [           main] c.e.c.m.c.ChaosDemoApplicationTests      : Starting ChaosDemoApplicationTests on WINX-19130 with PID 16448 (started by jisaew2 in C:\Users\jisaew2\code\external\chaos-monkey-spring-boot\demo-apps\chaos-monkey-web-reactive-app)
2019-11-13 14:08:49.870  INFO 16448 --- [           main] c.e.c.m.c.ChaosDemoApplicationTests      : The following profiles are active: chaos-monkey
2019-11-13 14:08:50.239  WARN 16448 --- [           main] r.c.GenericReactiveWebApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'meterRegistryPostProcessor' defined in class path resource [org/springframework/boot/actuate/autoconfigure/metrics/MetricsAutoConfiguration.class]: Initialization of bean failed; nested exception is java.lang.IllegalArgumentException: warning no match for this type name: org.springframework.data.repository.Repository [Xlint:invalidAbsoluteTypeName]
2019-11-13 14:08:50.243  INFO 16448 --- [           main] ConditionEvaluationReportLoggingListener : 

Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
2019-11-13 14:08:50.245 ERROR 16448 --- [           main] o.s.boot.SpringApplication               : Application run failed

org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'meterRegistryPostProcessor' defined in class path resource [org/springframework/boot/actuate/autoconfigure/metrics/MetricsAutoConfiguration.class]: Initialization of bean failed; nested exception is java.lang.IllegalArgumentException: warning no match for this type name: org.springframework.data.repository.Repository [Xlint:invalidAbsoluteTypeName]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:601) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:204) ~[spring-beans-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.context.support.PostProcessorRegistrationDelegate.registerBeanPostProcessors(PostProcessorRegistrationDelegate.java:240) ~[spring-context-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.registerBeanPostProcessors(AbstractApplicationContext.java:721) ~[spring-context-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:534) ~[spring-context-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:744) ~[spring-boot-2.1.8.RELEASE.jar:2.1.8.RELEASE]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:391) ~[spring-boot-2.1.8.RELEASE.jar:2.1.8.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:312) ~[spring-boot-2.1.8.RELEASE.jar:2.1.8.RELEASE]
	at org.springframework.boot.test.context.SpringBootContextLoader.loadContext(SpringBootContextLoader.java:120) ~[spring-boot-test-2.1.8.RELEASE.jar:2.1.8.RELEASE]
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:99) ~[spring-test-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:117) ~[spring-test-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.test.context.support.DefaultTestContext.getApplicationContext(DefaultTestContext.java:108) ~[spring-test-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.boot.test.autoconfigure.SpringBootDependencyInjectionTestExecutionListener.outputConditionEvaluationReport(SpringBootDependencyInjectionTestExecutionListener.java:53) ~[spring-boot-test-autoconfigure-2.1.8.RELEASE.jar:2.1.8.RELEASE]
	at org.springframework.boot.test.autoconfigure.SpringBootDependencyInjectionTestExecutionListener.prepareTestInstance(SpringBootDependencyInjectionTestExecutionListener.java:46) ~[spring-boot-test-autoconfigure-2.1.8.RELEASE.jar:2.1.8.RELEASE]
	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:246) ~[spring-test-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.createTest(SpringJUnit4ClassRunner.java:227) ~[spring-test-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner$1.runReflectiveCall(SpringJUnit4ClassRunner.java:289) ~[spring-test-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) ~[junit-4.12.jar:4.12]
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.methodBlock(SpringJUnit4ClassRunner.java:291) ~[spring-test-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:246) ~[spring-test-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97) ~[spring-test-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290) ~[junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71) ~[junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288) ~[junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58) ~[junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268) ~[junit-4.12.jar:4.12]
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61) ~[spring-test-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70) ~[spring-test-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363) ~[junit-4.12.jar:4.12]
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190) ~[spring-test-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365) ~[surefire-junit4-2.22.2.jar:2.22.2]
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273) ~[surefire-junit4-2.22.2.jar:2.22.2]
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238) ~[surefire-junit4-2.22.2.jar:2.22.2]
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159) ~[surefire-junit4-2.22.2.jar:2.22.2]
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384) ~[surefire-booter-2.22.2.jar:2.22.2]
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345) ~[surefire-booter-2.22.2.jar:2.22.2]
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126) ~[surefire-booter-2.22.2.jar:2.22.2]
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418) ~[surefire-booter-2.22.2.jar:2.22.2]
Caused by: java.lang.IllegalArgumentException: warning no match for this type name: org.springframework.data.repository.Repository [Xlint:invalidAbsoluteTypeName]
	at org.aspectj.weaver.tools.PointcutParser.parsePointcutExpression(PointcutParser.java:319) ~[aspectjweaver-1.9.4.jar:na]
	at org.springframework.aop.aspectj.AspectJExpressionPointcut.buildPointcutExpression(AspectJExpressionPointcut.java:227) ~[spring-aop-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.aop.aspectj.AspectJExpressionPointcut.obtainPointcutExpression(AspectJExpressionPointcut.java:198) ~[spring-aop-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.aop.aspectj.AspectJExpressionPointcut.getClassFilter(AspectJExpressionPointcut.java:177) ~[spring-aop-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.aop.support.AopUtils.canApply(AopUtils.java:225) ~[spring-aop-5.1.9.RELEASE.jar:5.1.9.RELEASE]
	at org.springframework.aop.support.AopUtils.canApply(AopUtils.java:288) ~[spring-aop-5.1.9.RELEASE.jar:5.1.9.RELEASE]
```

And here's the aspect as I have it coded right now:
```
public class SpringRepositoryAspect extends ChaosMonkeyBaseAspect {

    private final ChaosMonkeyRequestScope chaosMonkeyRequestScope;
    private MetricEventPublisher metricEventPublisher;
    private WatcherProperties watcherProperties;

    @Pointcut("this(org.springframework.data.repository.Repository) || within(@org.springframework.data.repository.RepositoryDefinition *)")
        public void isRepositoryImplementation() {
    }

    @Pointcut("within(@org.springframework.stereotype.Repository *)")
    public void hasRepositoryAnnotation() {
    }

    @Around("(hasRepositoryAnnotation() || isRepositoryImplementation()) && allPublicMethodPointcut() && !classInChaosMonkeyPackage()")
    public Object intercept(ProceedingJoinPoint pjp) throws Throwable {

```

And the ChaosMonkeyConfiguration:

```
    @Bean
    @DependsOn("chaosMonkeyRequestScope")
    public SpringRepositoryAspect repositoryAspect(ChaosMonkeyRequestScope chaosMonkeyRequestScope) {
        return new SpringRepositoryAspect(chaosMonkeyRequestScope, publisher(), watcherProperties);
    }
```

**Comment by maiksensi** [comment]
Wow you're fast. Thanks!
 That is strange, because we tested it here and it worked. I'll have a look at it as soon as I can (as I will be on a workshop tomorrow and the day after it is either the weekend or monday). If I get it to work, I'll push a commit to your fork if that's alright? One more thing, what java version / os youre using?

**Comment by ericwyles** [comment]
I've produce the issue on Java 11 (zulu-11) and Java 1.8 (openjdk-1.8.0_101) both on Windows 10.

Sure, feel free to push to that fork. Perfectly fine with me.

Thanks!

**Comment by ericwyles** [comment]
@maiksensi would it be helpful for me to go ahead and push the current state to my fork branch? I think it'll likely fail some tests in the PR validation process but it might be useful.

**Comment by maiksensi** [comment]
@ericwyles I am sorry that I didn't get to it yet, I will try my best tomorrow and let you know how it goes so we can finish up that thing here.

**Comment by maiksensi** [comment]
So, I basically adapted your approach, which seems the right one considering the current design. I tried a conditional aspect but that is not supported with spring as of now. I adapted the naming a bit in your and pushed some changes. Let me know if you think it is more clear this way @ericwyles and if so we can see that it gets merged soon.

**Comment by maiksensi** [comment]
fixed via #148
